### ISSUE TITLE
slowly consumes all the ram and crashes the esp8266!

### ISSUE THREAD

**Issue by RiccardoUva** [sinricpro/esp8266-esp32-sdk#195]
Hi, I've been working on a project for a few months now, recently I had to rewrite all the code so that it works with Sinric Pro as Sinric is closing in August... At the beginning I didn't understand why after a while my ESp8266 was crashing, this morning I took the DimSwitch example sketch and and modified it so that every 5 seconds it prints how much free ram is available...I noticed that it gradually fills it all and at the end the esp crashes, this didn't happen before with the codes from the old Sinric.... it always happens, even without sending commands even once! Always Exception (29).  How can i solve this issue now? I think it's a library problem... Thank you all in advance for the help!

 **CODE:**

```C++
// Uncomment the following line to enable serial debug output
//#define ENABLE_DEBUG

#ifdef ENABLE_DEBUG
#define DEBUG_ESP_PORT Serial
#define NODEBUG_WEBSOCKETS
#define NDEBUG
#endif

#include <Arduino.h>
#ifdef ESP8266
#include <ESP8266WiFi.h>
#endif
#ifdef ESP32
#include <WiFi.h>
#endif

#include "SinricPro.h"
#include "SinricProDimSwitch.h"

#define WIFI_SSID "SSID"
#define WIFI_PASS "PASS"
#define APP_KEY "API" // Should look like "de0bxxxx-1x3x-4x3x-ax2x-5dabxxxxxxxx"
#define APP_SECRET "APP SECRET" // Should look like "5f36xxxx-x3x7-4x3x-xexe-e86724a9xxxx-4c4axxxx-3x3x-x5xe-x9x3-333d65xxxxxx"
#define DIMSWITCH_ID "ID" // Should look like "5dc1564130xxxxxxxxxxxxxx"
#define BAUD_RATE 9600 // Change baudrate to your need

// we use a struct to store all states and values for our dimmable switch
struct {
  bool powerState = false;
  int powerLevel = 0;
} device_state;

unsigned long timer;

bool onPowerState(const String &deviceId, bool &state) {
  Serial.printf("Device %s power turned %s \r\n", deviceId.c_str(), state?"on":"off");
  device_state.powerState = state;
  return true; // request handled properly
}

bool onPowerLevel(const String &deviceId, int &powerLevel) {
  device_state.powerLevel = powerLevel;
  Serial.printf("Device %s power level changed to %d\r\n", deviceId.c_str(), device_state.powerLevel);
  return true;
}

bool onAdjustPowerLevel(const String &deviceId, int &levelDelta) {
  device_state.powerLevel += levelDelta;
  Serial.printf("Device %s power level changed about %i to %d\r\n", deviceId.c_str(), levelDelta, device_state.powerLevel);
  levelDelta = device_state.powerLevel;
  return true;
}

void setupWiFi() {
  Serial.printf("\r\n[Wifi]: Connecting");
  WiFi.begin(WIFI_SSID, WIFI_PASS);

  while (WiFi.status() != WL_CONNECTED) {
    Serial.printf(".");
    delay(250);
  }
  IPAddress localIP = WiFi.localIP();
  Serial.printf("connected!\r\n[WiFi]: IP-Address is %d.%d.%d.%d\r\n", localIP[0], localIP[1], localIP[2], localIP[3]);
}

void setupSinricPro() {
  SinricProDimSwitch &myDimSwitch = SinricPro[DIMSWITCH_ID];

  // set callback function to device
  myDimSwitch.onPowerState(onPowerState);
  myDimSwitch.onPowerLevel(onPowerLevel);
  myDimSwitch.onAdjustPowerLevel(onAdjustPowerLevel);

  // setup SinricPro
  SinricPro.onConnected([](){ Serial.printf("Connected to SinricPro\r\n"); });
  SinricPro.onDisconnected([](){ Serial.printf("Disconnected from SinricPro\r\n"); });
  SinricPro.begin(APP_KEY, APP_SECRET);
}

// main setup function
void setup() {
  Serial.begin(BAUD_RATE); Serial.printf("\r\n\r\n");
  setupWiFi();
  setupSinricPro();
}

void loop() {
  SinricPro.handle();
  //prints every 5 seconds the free ram
  if(millis() >= timer + 5000){
    timer= millis();
    Serial.println(ESP.getFreeHeap(),DEC);
  }

}
```

**SERIAL MONITOR**
<details><summary>Click to expand</summary>

```
[Wifi]: Connecting.....................connected!
[WiFi]: IP-Address is 192.168.188.68
18064
Connected to SinricPro
18336
17832
17808
17808
17808
17808
17808
17808
17784
17784
17784
17784
17784
17784
17760
17760
17760
17760
17760
17760
17760
17760
17760
17760
17760
17760
17760
17760
17760
17736
17736
17736
17712
17712
17712
17688
17688
17688
17664
17664
17664
17640
17640
17640
17640
17640
17640
17640
17640
17640
17640
17640
17640
17616
17616
17616
17568
17568
17568
17520
17520
17520
17472
17472
17472
17424
17424
17424
17400
17400
17400
17400
17400
17400
17400
17400
17400
17376
17376
17376
17304
17304
17304
17208
17208
17208
17112
17112
17112
17016
17016
17016
16944
16944
16944
16920
16920
16920
16920
16920
16920
16896
16728
16728
16632
16632
16632
16464
16464
16464
16272
16272
16272
16080
16080
16080
15912
15912
15912
15816
15816
15816
15792
15792
15792
15768
15768
15768
15648
15648
15648
15384
15384
15384
14840
14840
14840
14504
14504
14504
14144
14144
14144
13904
13904
13904
13784
13784
13784
13736
13736
13736
13592
13592
13592
13256
13256
13256
12848
12848
12848
12392
12392
12392
11840
11840
11840
11408
11408
11408
11168
11168
11168
11000
11000
11000
10808
10808
10808
10448
10448
10448
10040
10040
10040
9488
9488
9488
8768
8768
8768
8168
8168
8168
7688
7688
7688
7424
7424
7424
7064
7064
7064
6656
6656
6656
6248
6248
6248
5672
5672
5672
4904
4904
4904
4160
4160
4160
3560
3560
3560
2960
2960
2960
2504
2504
2504
1928
1928
1928
1496
1496
1496
968
968
968
584
584
584

--------------- CUT HERE FOR EXCEPTION DECODER ---------------

Exception (29):
epc1=0x4000df64 epc2=0x00000000 epc3=0x00000000 excvaddr=0x00000000 depc=0x00000000

>>>stack>>>

ctx: sys
sp: 3fffec10 end: 3fffffb0 offset: 0190
3fffeda0:  00000000 3ffee430 00000002 40100364  
3fffedb0:  40240c67 401050b7 3ffed3d0 40240bfc  
3fffedc0:  00000002 40240ba3 00000002 4023fcfc  
3fffedd0:  4023fd25 3fffee80 3ffee430 00000016  
3fffede0:  4023d78c 3fffee80 3ffee2d8 3ffedcdc  
3fffedf0:  3ffeaea4 3fffee80 3fffee80 40220048  
3fffee00:  65746552 646e5520 72677265 646e756f  
3fffee10:  00000000 3ffef618 000000ff 00000000  
3fffee20:  40239af1 3ffefa24 3fffbae4 40218b85  
3fffee30:  ffffffc5 3ffeddec 3ffeaeb4 3ffee430  
3fffee40:  3ffed3d0 00000025 00000000 4023e487  
3fffee50:  00000000 3ffefc24 ffffffc5 00000000  
3fffee60:  00000000 3ffee430 0012a219 0000000f  
3fffee70:  00000000 00000000 002afa15 40220b11  
3fffee80:  00000000 00310b0b 00640102 0000004f  
3fffee90:  3ffeaec8 000000ec 3ffeaee7 3ffeaebc  
3fffeea0:  3ffeaeed 3ffeaec8 3ffeaeda 3ffeaef8  
3fffeeb0:  00000000 00000000 3ffeaf7f 3ffeaf4c  
3fffeec0:  3ffeaefe 3ffeaf1a 00000000 3ffeaf95  
3fffeed0:  00000000 00000000 00000025 00000000  
3fffeee0:  3ffefecc 4023def6 3ffed3d0 3ffefc24  
3fffeef0:  00000000 3ffee430 3ffed3d0 3ffeaea4  
3fffef00:  3ffeaea4 00000114 00000000 00000025  
3fffef10:  00000000 3ffeaeae 40251643 3ffed3d0  
3fffef20:  3ffeae98 3fffdcc0 3ffe9a88 3ffe9a88  
3fffef30:  00000080 3ffed3d0 00000000 3ffe85dc  
3fffef40:  40250f07 3fffdab0 00000000 4020a8b2  
3fffef50:  3ffe9a88 40000f49 3fffdab0 40000f49  
3fffef60:  40000e19 00067931 00000000 00000005  
3fffef70:  00000000 aa55aa55 000000b5 40105199  
3fffef80:  4010519f 00000000 00000005 74616572  
3fffef90:  4010000d 76656400 49656369 65720064  
3fffefa0:  40245df8 3fffef3c 40245db1 3ffffe18  
3fffefb0:  3fffffc0 00000000 00000000 feefeffe  
3fffefc0:  feefeffe feefeffe feefeffe feefeffe  
3fffefd0:  feefeffe feefeffe feefeffe feefeffe  
3fffefe0:  feefeffe feefeffe feefeffe feefeffe  
3fffeff0:  feefeffe feefeffe feefeffe feefeffe  
3ffff000:  feefeffe feefeffe feefeffe feefeffe  
3ffff010:  feefeffe feefeffe feefeffe feefeffe  
3ffff020:  feefeffe feefeffe feefeffe feefeffe  
3ffff030:  feefeffe feefeffe feefeffe feefeffe  
3ffff040:  feefeffe feefeffe feefeffe feefeffe  
3ffff050:  feefeffe feefeffe feefeffe feefeffe  
3ffff060:  feefeffe feefeffe feefeffe feefeffe  
3ffff070:  feefeffe feefeffe feefeffe feefeffe  
3ffff080:  feefeffe feefeffe feefeffe feefeffe  
3ffff090:  feefeffe feefeffe feefeffe feefeffe  
3ffff0a0:  feefeffe feefeffe feefeffe feefeffe  
3ffff0b0:  feefeffe feefeffe feefeffe feefeffe  
3ffff0c0:  feefeffe feefeffe feefeffe feefeffe  
3ffff0d0:  feefeffe feefeffe feefeffe feefeffe  
3ffff0e0:  feefeffe feefeffe feefeffe feefeffe  
3ffff0f0:  feefeffe feefeffe feefeffe feefeffe  
3ffff100:  feefeffe feefeffe feefeffe feefeffe  
3ffff110:  feefeffe feefeffe feefeffe feefeffe  
3ffff120:  feefeffe feefeffe feefeffe feefeffe  
3ffff130:  feefeffe feefeffe feefeffe feefeffe  
3ffff140:  feefeffe feefeffe feefeffe feefeffe  
3ffff150:  feefeffe feefeffe feefeffe feefeffe  
3ffff160:  feefeffe feefeffe feefeffe feefeffe  
3ffff170:  feefeffe feefeffe feefeffe feefeffe  
3ffff180:  feefeffe feefeffe feefeffe feefeffe  
3ffff190:  feefeffe feefeffe feefeffe feefeffe  
3ffff1a0:  feefeffe feefeffe feefeffe feefeffe  
3ffff1b0:  feefeffe feefeffe feefeffe feefeffe  
3ffff1c0:  feefeffe feefeffe feefeffe feefeffe  
3ffff1d0:  feefeffe feefeffe feefeffe feefeffe  
3ffff1e0:  feefeffe feefeffe feefeffe feefeffe  
3ffff1f0:  feefeffe feefeffe feefeffe feefeffe  
3ffff200:  feefeffe feefeffe feefeffe feefeffe  
3ffff210:  feefeffe feefeffe feefeffe feefeffe  
3ffff220:  feefeffe feefeffe feefeffe feefeffe  
3ffff230:  feefeffe feefeffe feefeffe feefeffe  
3ffff240:  feefeffe feefeffe feefeffe feefeffe  
3ffff250:  feefeffe feefeffe feefeffe feefeffe  
3ffff260:  feefeffe feefeffe feefeffe feefeffe  
3ffff270:  feefeffe feefeffe feefeffe feefeffe  
3ffff280:  feefeffe feefeffe feefeffe feefeffe  
3ffff290:  feefeffe feefeffe feefeffe feefeffe  
3ffff2a0:  feefeffe feefeffe feefeffe feefeffe  
3ffff2b0:  feefeffe feefeffe feefeffe feefeffe  
3ffff2c0:  feefeffe feefeffe feefeffe feefeffe  
3ffff2d0:  feefeffe feefeffe feefeffe feefeffe  
3ffff2e0:  feefeffe feefeffe feefeffe feefeffe  
3ffff2f0:  feefeffe feefeffe feefeffe feefeffe  
3ffff300:  feefeffe feefeffe feefeffe feefeffe  
3ffff310:  feefeffe feefeffe feefeffe feefeffe  
3ffff320:  feefeffe feefeffe feefeffe feefeffe  
3ffff330:  feefeffe feefeffe feefeffe feefeffe  
3ffff340:  feefeffe feefeffe feefeffe feefeffe  
3ffff350:  feefeffe feefeffe feefeffe feefeffe  
3ffff360:  feefeffe feefeffe feefeffe feefeffe  
3ffff370:  feefeffe feefeffe feefeffe feefeffe  
3ffff380:  feefeffe feefeffe feefeffe feefeffe  
3ffff390:  feefeffe feefeffe feefeffe feefeffe  
3ffff3a0:  feefeffe feefeffe feefeffe feefeffe  
3ffff3b0:  feefeffe feefeffe feefeffe feefeffe  
3ffff3c0:  feefeffe feefeffe feefeffe feefeffe  
3ffff3d0:  feefeffe feefeffe feefeffe feefeffe  
3ffff3e0:  feefeffe feefeffe feefeffe feefeffe  
3ffff3f0:  feefeffe feefeffe feefeffe feefeffe  
3ffff400:  feefeffe feefeffe feefeffe feefeffe  
3ffff410:  feefeffe feefeffe feefeffe feefeffe  
3ffff420:  feefeffe feefeffe feefeffe feefeffe  
3ffff430:  feefeffe feefeffe feefeffe feefeffe  
3ffff440:  feefeffe feefeffe feefeffe feefeffe  
3ffff450:  feefeffe feefeffe feefeffe feefeffe  
3ffff460:  feefeffe feefeffe feefeffe feefeffe  
3ffff470:  feefeffe feefeffe feefeffe feefeffe  
3ffff480:  feefeffe feefeffe feefeffe feefeffe  
3ffff490:  feefeffe feefeffe feefeffe feefeffe  
3ffff4a0:  feefeffe feefeffe feefeffe feefeffe  
3ffff4b0:  feefeffe feefeffe feefeffe feefeffe  
3ffff4c0:  feefeffe feefeffe feefeffe feefeffe  
3ffff4d0:  feefeffe feefeffe feefeffe feefeffe  
3ffff4e0:  feefeffe feefeffe feefeffe feefeffe  
3ffff4f0:  feefeffe feefeffe feefeffe feefeffe  
3ffff500:  feefeffe feefeffe feefeffe feefeffe  
3ffff510:  feefeffe feefeffe feefeffe feefeffe  
3ffff520:  feefeffe feefeffe feefeffe feefeffe  
3ffff530:  feefeffe feefeffe feefeffe feefeffe  
3ffff540:  feefeffe feefeffe feefeffe feefeffe  
3ffff550:  feefeffe feefeffe feefeffe feefeffe  
3ffff560:  feefeffe feefeffe feefeffe feefeffe  
3ffff570:  feefeffe feefeffe feefeffe feefeffe  
3ffff580:  feefeffe feefeffe feefeffe feefeffe  
3ffff590:  feefeffe feefeffe feefeffe feefeffe  
3ffff5a0:  feefeffe feefeffe feefeffe feefeffe  
3ffff5b0:  feefeffe feefeffe feefeffe feefeffe  
3ffff5c0:  feefeffe feefeffe feefeffe feefeffe  
3ffff5d0:  feefeffe fee⸮:+⸮⸮Rf⸮`⸮
```

</details>

**Comment by sivar2311** [comment]
Please let me know the used versions of:
- ArduinoCore
- SinricPro
- ArduinoJson 
- WebSocket

Update to the latest available versions and run the test again.

Here is my log:
```
[Wifi]: Connecting.............connected!
[WiFi]: IP-Address is 192.168.2.73
17376
Connected to SinricPro
17656
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
17136
```

**Comment by kakopappa** [comment]
Hi,

Can you post the versions of the following libraries please

1. esp8266 sdk from boards manager
2. Websocket from library manager
3. ArduinoJson from library manager

**Comment by kakopappa** [comment]
If you get a chance please decode the exception using

https://github.com/me-no-dev/EspExceptionDecoder

And post it here

**Comment by RiccardoUva** [comment]
Hi! First of all thanks for the reply and your time! 

Unfortunately the esp restarts before printing the whole exception, yesterday it managed to print it completely, I decoded it and this is what came out:

Exception 29: StoreProhibited: A store referenced a page mapped with an attribute that does not permit stores
PC: 0x4000df64
EXCVADDR: 0x00000000

0x40100574: pvPortZalloc(size_t, char const*, int) at C:\Users\Riccardo\AppData\Local\Arduino15\packages\esp8266\hardware\esp8266\3.0.0\cores\esp8266\umm_malloc/umm_heap_select.h line 85
------------------------------------------------------------------------------
These are the versions that i'm using.
Esp8266 SDK 3.0.0
Websockets 2.3.5
ArduinoJson 6.18.0
Arduino IDE 1.8.15

Thanks for the help!

**Comment by RiccardoUva** [comment]
> Please let me know the used versions of:
> 
> * ArduinoCore
> * SinricPro
> * ArduinoJson
> * WebSocket
> 
> Update to the latest available versions and run the test again.
> 
> Here is my log:
> 
> ```
> [Wifi]: Connecting.............connected!
> [WiFi]: IP-Address is 192.168.2.73
> 17376
> Connected to SinricPro
> 17656
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> 17136
> ```

**Comment by sivar2311** [comment]
I am not able to reproduce this issue.
I recommend to uninstall your Arduino installation completely and delete all remaining files . Then do a clean reinstallation.

If this error is really present in the SDK, then many users must have already reported this error.

Here is my setup:
![image](https://user-images.githubusercontent.com/2558564/123159593-7be41100-d46d-11eb-9392-cc70f45207ef.png)

And here is my log (running the sketch you provided above):
```
[Wifi]: Connecting.............connected!
[WiFi]: IP-Address is 192.168.2.73
18056
Connected to SinricPro
18328
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17824
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
17656
```

**Comment by RiccardoUva** [comment]
Hi, sorry for the absence of news but I've been busy ... anyway back to the discussion, after reinstalling everything and made a clean installation the problem continued to occur ... at home I have two wifi networks, one in cascade with the other, controlled by two different routers ... until now it was always connected to the second wifi router (the second wifi network), connecting the esp to the primary router the problem solves... I do not know exactly what it is, but anyway the problem for me in a sense has been resolved. I hope I've been helpful if the same thing happens to someone in the future.

_**TLDR: Try connecting the esp to another wifi network.**_

Thank you all for your help and have a good time!

**Comment by sivar2311** [comment]
Thank you for the feedback.
My ESP is also on an extra WiFi network behind a second router. But I have not encountered this problem yet.
Then it seems to be a problem with the ESP core or the ESP Arduino core.
